### ISSUE TITLE
Cannot use callback function parameter of jsPDF

### ISSUE THREAD

**Issue by Colinh84** [eKoopmans/html2pdf#5]
When using jsPDF on it's own, you can use a callback function parameter when initializing to perform tasks once the PDF is created. With html2pdf, there is no way to pass in this callback function. I've had to resort to using a timer to do some DOM cleanup tasks after PDF generation that would normally be put in a callback. It would be nice if we could pass a callback function into the jsPDF object options parameter when calling html2pdf.

**Comment by eKoopmans** [comment]
Hi @Colinh84, thanks for the feedback! Can you give me a code example of how you would provide a callback to `jsPDF`, and/or a link to the relevant docs? Thanks!

**Comment by Colinh84** [comment]
Hi @eKoopmans, I was actually mistaken and it was the old addHTML() function in jsPDF and not the calling of jsPDF itself which allowed the callback. I still think it would be extremely beneficial to allow an optional callback function when calling html2pdf, or better yet, turn html2pdf into a promise so it can be called as:

```javascript
html2pdf(doc, options).then(function(pdf){
//do custom stuff after pdf generation
});
```

**Comment by eKoopmans** [comment]
Right, I agree. I'm planning a revamp to provide a broader API, I'll make sure that's included. I'm all for promises.

**Comment by ssougnez** [comment]
Hi, do you have an update on this please ? I'm using html2pdf in production right now but it's tricky to display something like "Generating..." as I have no idea when the generation will be done and the time used depends on the page size.

This seems to be a pretty critical functionality, so maybe you should plan a release with only this?

Thanks a lot for the good work

**Comment by sharmarohit** [comment]
i second comments by ssougnez, i think having a callback in very important. I need similar functionality where I need to display generating pdf.

Thanks for all the good work.

**Comment by eKoopmans** [comment]
Hi guys, sorry to keep you hanging. Yes I'm actively working on the new API, and it's getting close to ready for testing, but if you need something that will work right now it's probably best to go with the pdf-callback branch described [here](https://github.com/eKoopmans/html2pdf/issues/14#issuecomment-318798955). It's been a long time coming, but life tends to get in the way.

I'll post here when the new API is ready for testing.

**Comment by lcrojanouninorte** [comment]
same 👍

**Comment by eKoopmans** [comment]
Hi, I have just pushed a new branch, [new-api](https://github.com/eKoopmans/html2pdf/tree/new-api), which is ready for testing! [The readme](https://github.com/eKoopmans/html2pdf/tree/new-api#advanced-usage) has all the information on the new API. Please let me know if there are any problems with it - I want to make sure it's stable before replacing the master!

**Comment by ssougnez** [comment]
Hi,

so the `save` function will return a promise that will resolve once the file is proposed to the user for download?

I just read your doc, it seems awesome, I'll test that on Monday on my work project. Clients will be happy ah ah ah

Just another off-topic question: do you follow releases of html2canvas closely ? It seems they are working on a v1.0.0 (alpha-10 currently) that includes a lot of new functionality that I need, so I was wondering if you were planning on releasing a new version of html2pdf including the new version (when it's released).

Thanks a lot for your awesome work :-D

**Comment by eKoopmans** [comment]
Hey @ssougnez, yes the `.save` method (and every other method) returns a promise that will resolve when done.

And yes, the html2canvas dev has been doing great work, it's probably stable enough now to switch over. I'll look into it more closely soon.

**Comment by ssougnez** [comment]
I never tested a different branch than master so far, how can I install the new-api branch with npm?

**Comment by eKoopmans** [comment]
Good question! Two options:

1. [Install from branch](https://stackoverflow.com/a/39732501/4080966): `npm install git://github.com/eKoopmans/html2pdf.git#new-api --save`

2. [Install from tarball](https://stackoverflow.com/a/37295723/4080966): `npm install https://github.com/eKoopmans/html2pdf/tarball/new-api`
(I've found this to be faster than above)

**Comment by ssougnez** [comment]
Hi,

I'm back with results. I could install the new-api branch via the tarball. At first glance, it seems to work pretty well. My PDF is still generated and even though there is a slight delay between the resolving of the promise and the document download proposal by the browser, this delay is very acceptable.

Of course, my usage of html2pdf is pretty limited. I get a reference to a DOM element, clone it, add some new HTML element in the clone, then:

`html2pdf().from(clone).set(options).save()`

And when I `await` (I'm using TypeScript) on that, it works as it should. The PDF in itself is the same as the one generated by the master version so, for me, all the lights are green.

One tiny remark thought, in the readme file of the new branch, there is a type in your example. You create the `opt` object but the closing tag is `})` instead of `}`.

Great job you did there ;-)

**Comment by eKoopmans** [comment]
Thanks @ssougnez for testing and for catching the mistake in the readme! Interesting about the delay between resolution and file prompt - `jsPDF.save` doesn't return a promise, so `html2pdf.save` resolves when save is *invoked*, not when it finishes.

It's not all green lights unfortunately for the new API, I'm getting problems with Promise polyfills over at #14. Let me know if any other issues crop up!

**Comment by eKoopmans** [comment]
P.S. You can also use `.then` with the old-style syntax, if you're not doing anything else fancy:

```js
html2pdf(element, opt).then(...)
```

**Comment by maxshuty** [comment]
@eKoopmans I'm having some trouble getting this to work. Is the callback supposed to be used like this?

```html2pdf().from(quoteEl).then(function(pdf) { 
        //pdf is null when I log this...
       console.log(pdf);
   }).save();
```
I'm currently on v.0.9.0. All that I really need to get is the base64 so I can send that back to the server and attach it to an email - it doesn't really matter to me how I accomplish that but I'm having issues figuring out how to use this callback properly.

Thanks in advance for any guidance.

**Comment by Greensahil** [comment]
maxshuty please tell me how you managed to send the btoa back to the server.

**Comment by maxshuty** [comment]
@Greensahil

Two things:
1) don’t do this unless you have a way of sanitizing the PDF on the server, otherwise you could potentially (and easily) have a user send very harmful things back to your server.

2) If you still insist you want to do this: just take the code I posted above and take the output of what I am console logging and throw it into an AJAX POST.

**Comment by eKoopmans** [comment]
@maxshuty glad you resolved your issue, sorry I wasn't able to help you with it. I'll see if I can make the documentation a bit clearer on how to do that sort of task.

@Greensahil unless you need to generate your PDFs on client-side, I strongly recommend using a server-side solution - you can setup [wkhtmltopdf](https://wkhtmltopdf.org/) on your server and call it using PHP/etc, it will produce better results than html2pdf.

I'll also close this issue, I believe it is fully resolved with v0.9.0.
